### PR TITLE
Improve/mdc

### PR DESCRIPTION
**Pull request by testower**
Related to https://github.com/entur/lamassu/issues/250

This adds possibility to provide interceptor for subscription updates. Example implementation from application code:

```java
public class LoggingSubscriptionUpdateInterceptor implements SubscriptionUpdateInterceptor {
    private final FeedProvider feedProvider;

    public LoggingSubscriptionUpdateInterceptor(
            FeedProvider feedProvider
    ) {
        this.feedProvider = feedProvider;
    }
    @Override
    public void beforeUpdate() {
        MDC.put("systemId", feedProvider.getSystemId());
    }

    @Override
    public void afterUpdate() {
        MDC.remove("systemId");
    }
}
```

Which in turn enables systemId to be logged from log statements within this library by configuring logback as this example shows:

```xml
            <appender name="console" class="ch.qos.logback.core.ConsoleAppender">
                <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
                    <providers>
                        <pattern>
                            <pattern>
                                {
                                    "serviceContext": {
                                        "service": "lamassu"
                                    },
                                    "mdc": {
                                        "systemId": "%X{systemId}"
                                    },
                                    "message": "%message\n%ex{full}",
                                    "severity": "%level",
                                    "reportLocation": {
                                        "filePath": "%logger",
                                        "lineNumber": "%line",
                                        "functionName": "%method"
                                    }
                                }
                            </pattern>
                        </pattern>
                    </providers>
                </encoder>
            </appender>
```

### CODE REVIEW

**Comment by testower** [comment]
I wonder if there's some sort of best practice for adding stuff to MDC from libraries. Maybe namespacing?

**Comment by testower** [comment]
These may also be relevant:

https://stackoverflow.com/questions/6073019/how-to-use-mdc-with-thread-pools
https://talktotheduck.dev/logging-best-practices-mdc-ingestion-and-scale#heading-plan-for-failure-eg-custom-thread-pool

**Comment by testower** [comment]
One way to improve this is to add some lifecycle hooks or interceptors to the subscription, so that the application can make the MDC calls in stead of the library code.

File: src/main/java/org/entur/gbfs/SubscriptionUpdateInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.entur.gbfs;
+
+public interface SubscriptionUpdateInterceptor {
```
Comment (testower) [inline src/main/java/org/entur/gbfs/SubscriptionUpdateInterceptor.java:3]: document

File: src/main/java/org/entur/gbfs/GbfsSubscriptionManager.java [hunk 2]
```diff
@@ -80,16 +100,35 @@ public String subscribeV3(
     return subscribe(new GbfsV3Subscription(options, consumer));
   }
 
+  /**
+   * Start a subscription on a GBFS v3.x feed
+   *
+   * @param options Options
+   * @param consumer A consumer that will handle receiving updates from the loader}
+   * @param updateInterceptor
```
Comment (testower) [inline src/main/java/org/entur/gbfs/GbfsSubscriptionManager.java:108]: description

File: src/main/java/org/entur/gbfs/GbfsSubscriptionManager.java [hunk 3]
```diff
@@ -66,6 +66,26 @@ public String subscribeV2(
...
+  /**
+   * Start a subscription on a GBFS v2.x feed
+   * <p>
+   * Since v2.x is backwards-compatible with v1.x, v1.x feeds can also be
+   * consumed with this subscription.
+   * </p>
+   *
+   * @param options Options
+   * @param consumer A consumer that will handle receiving updates from the loader
+   * @param updateInterceptor
```
Comment (testower) [inline src/main/java/org/entur/gbfs/GbfsSubscriptionManager.java:78]: description

File: src/main/java/org/entur/gbfs/loader/v2/GbfsV2Subscription.java [hunk 4]
```diff
@@ -108,6 +125,10 @@ public void update() {
       );
       consumer.accept(delivery);
     }
+
```
Comment (testower) [inline src/main/java/org/entur/gbfs/loader/v2/GbfsV2Subscription.java:128]: Put this afterUpdate in a finally block for a try block around the update call (same for v3)

**Review by hbruch — CHANGES_REQUESTED** [review]
Calling the `updateInterceptor` in the `update` method does not support the intended use case to have a the system_id logged in case of exceptions, as exception logging occurs after `update()` was called and hence the externally set MDC is cleared already.

Possible solution could be to add `beforeUpdate`/`afterUpdate` as lifecycle methods to the GBFSSubscription classes and let them be called by the GBFSSubscriptionManager before/after calling update, e.g. like this
```java
public void update() {
    subscriptions
      .values()
      .parallelStream()
      .forEach(subscription ->
        Optional
          .ofNullable(customThreadPool)
          .orElse(ForkJoinPool.commonPool())
          .execute(() -> this.update(subscription))
      );
  }

  public void update(GbfsSubscription subscription) {
    try {
      subscription.beforeUpdate();
      subscription.update();
    } catch (RuntimeException e) {
      LOG.error("Error updating subscription", e);
    } finally {
      subscription.afterUpdate();
    }
  }
```

**Pull request merged**